### PR TITLE
feat: add Dynamic Fine-Tuning (DFT) loss support

### PR DESCRIPTION
**Pull request by kryvokhyzha**
# Description

Add support for **Dynamic Fine-Tuning (DFT)** loss.

Based on the paper [*On the Generalization of Supervised Fine-Tuning: A Reinforcement Learning
Perspective with Reward Rectification*](https://arxiv.org/abs/2508.05629), DFT rescales the per-token cross-entropy loss by the model's own predicted probability (with stop-gradient):

```
dft_loss_t = p(y_t).detach() * (-log p(y_t))
```

where standard SFT uses `sft_loss_t = -log p(y_t)`.

# Tests

6 unit tests added and passing:

| Test | Description |
|------|-------------|
| `test_dft_loss_less_than_or_equal_standard` | DFT ≤ standard CE since p(y_t) ≤ 1 |
| `test_dft_loss_non_negative` | DFT loss is non-negative |
| `test_dft_equals_standard_when_confident` | Both losses ≈ 0 when model is confident |
| `test_dft_down_weights_uncertain_tokens` | Ratio matches 1/V for uniform logits |
| `test_dft_gradient_differs_from_standard` | DFT produces different gradients than SFT |
| `test_dft_stop_gradient_on_probability` | Confirms stop_gradient on p(y_t) matters |

```
tests/unit/dft_loss_test.py::DFTLossTest::test_dft_down_weights_uncertain_tokens PASSED
tests/unit/dft_loss_test.py::DFTLossTest::test_dft_equals_standard_when_confident PASSED
tests/unit/dft_loss_test.py::DFTLossTest::test_dft_gradient_differs_from_standard PASSED
tests/unit/dft_loss_test.py::DFTLossTest::test_dft_loss_less_than_or_equal_standard PASSED
tests/unit/dft_loss_test.py::DFTLossTest::test_dft_loss_non_negative PASSED
tests/unit/dft_loss_test.py::DFTLossTest::test_dft_stop_gradient_on_probability PASSED
======================== 6 passed in 12.83s ========================
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by shralex** [comment]
Hi @kryvokhyzha thank you for your contribution! While initially MaxText supported SFT natively (and we're still using this in some parts of the code), we've migrated to use Tunix for SFT and other post-training methods. So your contribution would be better in Tunix, and then you can integrate it in MaxText. 

Here's the trainer using Tunix SFT in MaxText:

https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/trainers/post_train/sft/train_sft.py

**Comment by kryvokhyzha** [comment]
Hi @shralex! Thanks, I'll check `tunix` and then reopen this issue
